docs/adrs/0549-composite-run-steps.md

```diff
@@ -0,0 +1,321 @@
+```
+
+Example `user/composite/action.yml`
+
+```yaml
+runs:
+  using: "composite"
+  steps:
+    - run: pip install -r requirements.txt
+    - run: npm install
+```
+
+Example Output
+
+```yaml
+[npm installation output]
+[pip requirements output]
+echo hello world 3
+echo hello world 4
+```
+
+We add a token called "composite" which allows our Runner code to process composite actions. By invoking "using: composite", our Runner code then processes the "steps" attribute, converts this template code to a list of steps, and finally runs each run step sequentially. If any step fails and there are no `if` conditions defined, the whole composite action job fails. 
+
+## Inputs
+
+Example `workflow.yml`:
+
+```yaml
+steps: 
+  - id: foo
+    uses: user/composite@v1
+    with:
+      your_name: "Octocat"
+```
+
+Example `user/composite/action.yml`:
+
+```yaml
+inputs:
+  your_name:
+    description: 'Your name'
+    default: 'Ethan'
+runs:
+  using: "composite"
+  steps: 
+    - run: echo hello ${{ inputs.your_name }}
+```
+
+Example Output:
+
+```
+hello Octocat
+```
+
+Each input variable in the composite action is only viewable in its own scope.
+
+## Outputs
+
+Example `workflow.yml`:
+
+```yaml
+...
+steps: 
+  - id: foo
+    uses: user/composite@v1
+  - run: echo random-number ${{ steps.foo.outputs.random-number }} 
+```
+
+Example `user/composite/action.yml`:
+
+```yaml
+outputs:
+  random-number: 
+    description: "random number"
+runs:
+  using: "composite"
+  steps: 
+    - run: echo "::set-output name=random-number::$(echo $RANDOM)"
+```
+
+Example Output:
+
+```
+::set-output name=my-output::43243
+random-number 43243
+```
+
+Each of the output variables from the composite action is viewable from the workflow file that uses the composite action. In other words, every child action output(s) is viewable only by its parent using dot notation (ex `steps.foo.outputs.bar`).
+
+Moreover, the output ids are only accessible within the scope where it was defined. Note that in the example above, in our `workflow.yml` file, it should not have access to output id (i.e. `my-output`). For example, in the `workflow.yml` file, you can't run `foo.steps.my-step.my-output`.
+
+## Context
+
+Similar to the workflow file, the composite action has access to the [same context objects](https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#contexts) (ex: `github`, `env`, `strategy`). 
+
+## Environment
+
+Example `workflow.yml`:
+
+```yaml
+env:
+  NAME1: test1
+  SERVER: production
+steps: 
+  - id: foo
+    uses: user/test@v1
+  - run: echo Server $SERVER
+```
+
+Example `user/composite/action.yml`:
+
+```yaml
+using: 'composite' 
+env:
+  NAME2: test2
+  SERVER: development
+runs:
+  using: "composite"
+  steps: 
+    - id: my-step
+      run: |
+        echo NAME2 $NAME2
+        echo Server $SERVER 
+      env:
+        NAME2: test3
+```
+
+Example Output:
+
+```
+NAME2 test3
+Server development
```

Comment:
According to this logic, workflow authors won't be able to override environment variables.
On the one hand, I can see how it would be useful for an action author to want full control over their scope, to avoid collisions (I can't think of an example off-hand). 
On the other, I can see it as useful for the action author to set a default, with the intention of it being overwritten (all our deploy scripts use the environment variable `STAGE` to control which environment to manage).

A good compromise here might be to allow the use of `input` within the top level `env`.

Reply:
For additional clarity in the example, can we also explicitly say what happens when you specify `env` on the step itself. 
My assumption is that it will behave the same as above, though it could be a bit of a surprise to workflow authors when explicitly setting an env doesn't change what the action has set